[FuncSpec] Improve accounting of specialization codesize growth

hazzlim opened the pull request:
Only accumulate the codesize increase of functions that are actually
specialized, rather than for every candidate specialization that we
analyse.

This fixes a subtle bug where prior analysis of candidate
specializations that were deemed unprofitable could prevent subsequent
profitable candidates from being recognised.

hazzlim committed:
[FuncSpec] Improve accounting of specialization codesize growth

Only accumulate the codesize increase of functions that are actually
specialized, rather than for every candidate specialization that we
analyse.

This fixes a subtle bug where prior analysis of candidate
specializations that were deemed unprofitable could prevent subsequent
profitable candidates from being recognised.

hazzlim committed:
[FuncSpec] Add tests for MaxCodeSizeGrowth

hazzlim committed:
Update tests

labrinea commented on llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:
```diff
@@ -924,16 +932,14 @@ bool FunctionSpecializer::findSpecializations(Function *F, unsigned FuncSize,
 
+      unsigned CodeSizeSavings = getCostValue(CodeSize);
+      unsigned CodeSizeCost = FuncSize - CodeSizeSavings;
+
```
I think we can sink these two back in the lambda where the TODO note was. I would rename CodeSizeCost to SpecSize.

labrinea replied:
Nevermind, I was wrong since it's being used outside of the lambda.

labrinea commented on llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:
```diff
@@ -964,7 +970,7 @@ bool FunctionSpecializer::findSpecializations(Function *F, unsigned FuncSize,
-        if (FunctionGrowth[F] / FuncSize > MaxCodeSizeGrowth)
+        if ((FunctionGrowth[F] + CodeSizeCost) / FuncSize > MaxCodeSizeGrowth)
           return false;
 
```
Here we successfully got past the FuncGrowth check. We know for sure the specialization will be created. We can safely do `FunctionGrowth[F] += SpecSize;` at this point, before returning `true`.

labrinea commented on llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:
```diff
@@ -964,7 +970,7 @@ bool FunctionSpecializer::findSpecializations(Function *F, unsigned FuncSize,
           return false;
         // Maximum codesize growth.
-        if (FunctionGrowth[F] / FuncSize > MaxCodeSizeGrowth)
+        if ((FunctionGrowth[F] + CodeSizeCost) / FuncSize > MaxCodeSizeGrowth)
```
s/CodeSizeCost/SpecSize

hazzlim replied:
Done

hazzlim commented on llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:
```diff
@@ -759,6 +771,14 @@ bool FunctionSpecializer::run() {
+    // codesize growth.
+    unsigned FuncSize = getCostValue(FunctionMetrics[S.F].NumInsts);
+    if ((FunctionGrowth[S.F] + S.CodeSizeCost) / FuncSize > MaxCodeSizeGrowth)
+      continue;
```
Yes, I think that is definitely the trade-off with moving the logic to here - we may skip some specializations, and because we do so after we have built the priority queue of specializations to perform, we may fail to specialize a "next-best" specialization candidate for a function that would have fit into the FunctionGrowth for a its function because it's place has been taken in the NSpecs best candidates.

However, if we move it to where you suggest we are making a slightly different tradeoff:
https://github.com/llvm/llvm-project/pull/113448#discussion_r1814682046
> Here we successfully got past the FuncGrowth check. We know for sure the specialization will be created. We can safely do FunctionGrowth[F] += SpecSize; at this point, before returning true.

If I understand correctly, we do not know for sure the specialization will be created, because we do not know that it will have a high enough score to end up in the NSpecs we decide to specialize for the module. If for a given function F we analyze N specialization candidates that are all deemed profitable and these N candidates use up the maximum FunctionGrowth for F, then we would deem the N+1-th candidate unprofitable based on FunctionGrowth even if it had a higher score (Inlining/CodeSize/Latency) than the candidates analyzed beforehand.

Maybe you still think this is the better tradeoff? 

I suppose to some extent it might make more sense to instead have a codesize growth across functions (per module) budget, and to build the priority queue of BestSpecs based on that rather than a budget based on number of specializations.

labrinea replied:
I think it is worth mentioning the background story. We initially came up with the FuncGrowth threshold to limit the overall codesize growth from being linear to the number of times the specializer runs. At first it NumIters was set to 1, and we increased it to 10 for enabling specialization of recursive functions. Imagine runnning 10 times yielded 10 times the original function size.

We have MaxClones to limit the amount of specializations per iteration. Within a single iteration of run, the growth has no effect basically unless you have many candidates (not the case for functions relying on promoteConstantStackValues). It is used to prevent the next iteration candidates. How can there be a N+1-th candidate if at the N-th step we decided to stop? (I am thinking of recursive functions).

Maybe we can keep `FunctionGrowth[S.F] += S.CodeSizeCost;` here but remove the early exit from the line above this comment. It should only remain in the isProfitable lambda in my opinion.

hazzlim replied:
Ah ok that makes a lot of sense, thank you for providing the additional context.

I have updated the code so that we only check the codesize increase inside the `isProfitable` lambda, but do the accumulation into `FunctionGrowth[S.F]` at the point we are actually creating the specialization. This has the downside of us potentially violating `MaxCodeSizeGrowth` in a single iteration, in the case where we have specialization candidates within a single iteration that individually do not cause us to exceed `MaxCodeSizeGrowth` (and so pass the check in the `isProfitable` lambda), but in combination do exceed when they are all performed. This is why I have removed the second test case from the regression test, as it no longer makes sense.

However I think that this is an acceptable trade-off, as it still preserves the intention of the MaxCodeSizeGrowth threshold by preventing linear codesize increase across multiple iterations, as even if the threshold is exceeded in an iteration we would not consider further specializations in successive iterations once the growth has been accounted for, and we have MaxClones to limit the amount of specializations per iteration as you say.

We could fix the potential violation of `MaxCodeSizeGrowth` within a single iteration by moving the accumulation into `FunctionGrowth[S.F]` to the end of the `isProfitable` lambda as you suggested. However I don't really like this, because we may end up discounting the most profitable specialization candidate for a function. To clarify my previous comment - by Nth and N+1-th candidates I meant candidates within a single iteration. If we accumulate the codesize growth whilst we are analyzing the candidates, we may 'use-up' MaxCodeSizeGrowth before encountering the best specialization and therefore fail to consider it:

```
// (Within a single iteration, MaxCodeSizeGrowth=100%)
spec1 [Score 500, Growth 50%]
spec2 [Score 500, Growth 50%]
// After examining the above, FunctionGrowth[S.F] = 100% so no more candidates will be considered profitable!
// Candidate with highest Score:
spec3 [Score 1000, Growth 10%]
```

labrinea commented on llvm/include/llvm/Transforms/IPO/FunctionSpecialization.h:
```diff
@@ -131,13 +131,16 @@ struct Spec {
   unsigned Score;
 
+  // Cost of the specialization, in terms of codesize.
+  unsigned CodeSizeCost;
```
If we decide to keep this field please s/CodeSizeCost/CodeSize and update the comment please ("Number of instructions of the specialization" or similar)

hazzlim replied:
Done

hazzlim committed:
Address review comments

- Rename local variable (s/CodeSizeCost/SpecSize)
- Rename struct member (s/CodeSizeCost/CodeSize) and update comment
- Remove early exit from specialization creation loop
- Remove redundant prefix from regression test name
- Remove regression test for exceeding FuncGrowth
